_episodes/l2_02_basic_unit_testing.md

```diff
@@ -0,0 +1,293 @@
+
+   
+
+2. **Consistent Formatting**: Consistency in naming conventions helps maintain a uniform and predictable test suite. Choose a naming style and stick to it throughout your test methods. Some common conventions include using **CamelCase** or **underscore-separated** words. Additionally, consider using a prefix like `test_` or a suffix like `should` to distinguish test methods from regular code.
+
+   
+
+3. **Single Responsibility**: Each test method should focus on testing a single aspect or behaviour of the unit under test. Avoid testing multiple scenarios within a single test method, as it can make the test less readable and harder to diagnose when failures occur. Instead, break down complex scenarios into multiple smaller tests, each targeting a specific case or condition.
+
+   
+
+4. **Test Data Organization**: Separate the test data from the test methods. Consider using dedicated variables or data structures to store test data. This allows for better readability and maintainability, as changes to the test data can be easily managed without modifying the test methods themselves.
+
+
+
+## 7. Unit Test for the Fibonacci Sequence
+
+With the above guidelines in mind, we can write some of the unit tests for our `Fibonacci.cpp` that we learnt in Chapter 1. The complete code is given below and can be found in [Chapter2](../code/Chapter2/Fibonacci_test.cpp).
+
+```cpp
+#include <iostream>
+#include "gtest/gtest.h"
+
+/*Returns the n'th term of the Fibonacci sequence.*/
+int recursive_fibonacci(int n)
+{
+if (n <= 1) {
+   return n;
+} else {
+   return recursive_fibonacci(n - 1) + recursive_fibonacci(n - 2);
+}
+}
+
+TEST(FibonacciTest, HandlesZeroInput) {
+EXPECT_EQ(recursive_fibonacci(0), 0);
+}
+
+TEST(FibonacciTest, HandlesValueOneAsInput) {
+EXPECT_EQ(recursive_fibonacci(1), 1);
+}
+
+TEST(FibonacciTest, HandlesPositiveInput) {
+EXPECT_EQ(recursive_fibonacci(5), 5);
+}
+
+int main(int argc, char **argv)
+{
+testing::InitGoogleTest(&argc, argv);
+return RUN_ALL_TESTS();
+}
+```
+
+On compiling and running above program, we get the following output.
+
+```bash
+[==========] Running 3 tests from 1 test suite.
+[----------] Global test environment set-up.
+[----------] 3 tests from FibonacciTest
+[ RUN      ] FibonacciTest.HandlesZeroInput
+[       OK ] FibonacciTest.HandlesZeroInput (0 ms)
+[ RUN      ] FibonacciTest.HandlesValueOneAsInput
+[       OK ] FibonacciTest.HandlesValueOneAsInput (0 ms)
+[ RUN      ] FibonacciTest.HandlesPositiveInput
+[       OK ] FibonacciTest.HandlesPositiveInput (0 ms)
+[----------] 3 tests from FibonacciTest (0 ms total)
+
+[----------] Global test environment tear-down
+[==========] 3 tests from 1 test suite ran. (0 ms total)
+[  PASSED  ] 3 tests.
+```
+
+
+
+### 7.1 Exercise
+
+Modify the above program to test for the negative values.
+
+**Solution**
+
+Add the following lines in your code.
+
+```cpp
+TEST(FibonacciTest, HandlesNegativeInput) {
+    EXPECT_EQ(recursive_fibonacci(-3), -3);
+}
+
+```
```

Comment:
I'm not sure about this as an exercise, because the result for putting -3 into the Fibonacci function is undefined. Yes, at the moment our function *will* return a value, but that's a shortcoming of the way its written. The reason for talking about trying to run the function with negative values in the last chapter was to get students to think about edge cases.

A better exercise might be to get the students to modify the function to throw an exception if negative values are input and then test for this.

Reply:
Resolved in d629e60